File: python/course-schedule-ii.py
from collections import defaultdict
from typing import List

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indeg = [0]*numCourses
        adj_list = defaultdict(set)
        
        for child, parent in prerequisites:
            adj_list[parent].add(child)
            indeg[child] += 1
        
        res = []
        stack = []
        
        for i, el in enumerate(indeg):
            if el == 0:
                stack.append(i)
                res.append(i)
        
        while stack:
            node = stack.pop()
            for nei in adj_list[node]:
                indeg[nei] -= 1
                if indeg[nei] == 0:
                    stack.append(nei)
                    res.append(nei)
        
        if len(res) == numCourses: return res
        else: return []

File: go/course_schedule_ii.go
package leetcode

func findOrder(numCourses int, prerequisites [][]int) []int {
	indegree := make([]int, numCourses)
	adj_list := make(map[int][]int)

	for _, el := range prerequisites {
		course, pre := el[0], el[1]

		indegree[course] += 1
		if _, ok := adj_list[course]; ok == false {
			adj_list[course] = make([]int, 0, 1)
		}

		adj_list[pre] = append(adj_list[pre], course)
	}

	stack := make([]int, 0)
	res := make([]int, 0)

	for i, el := range indegree {
		if el == 0 {
			stack = append(stack, i)
		}
	}

	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		for _, el := range adj_list[node] {
			indegree[el] -= 1
			if indegree[el] == 0 {
				stack = append(stack, el)
			}
		}

		res = append(res, node)
	}

	if len(res) == numCourses {
		return res
	} else {
		return make([]int, 0)
	}
}
